### PR TITLE
Add prettier, expose functionality to create libs and components

### PR DESCRIPTION
**Pull request by eliperelman**

### CODE REVIEW

**Comment by eliperelman** [comment]
I know this is mega, but I think it can be super helpful for everyone.

@edmorley feel free to have a browse and tell me if you think.

File: README.md [hunk 1]
```diff
@@ -60,7 +61,13 @@ by Neutrino. This means that by default all project source code should live in a
...
+| Project Type | Source | Output |
+| --- | --- | --- |
+| React app | `src` | `build` |
+| Web Library | `src` | `lib` |
+| React components | `src/components` | `lib` |
+
+## Quickstart React App
 
 After installing Neutrino and the this preset, add a new directory named `src` in the root of the project, with
 a single JS file named `index.js` in it.
```
Comment (helfi92) [inline README.md:73]: Since we're dealing with a React app, perhaps it would be useful to mention to have `react` and `react-dom` installed.

File: README.md [hunk 2]
```diff
@@ -185,6 +191,75 @@ module.exports = {
...
+      components: true
+    }]
+  ]
+};
+```
+
+## Building libraries and components
+
+When building libraries and components, they are output to the `lib` directory by default when running `neutrino build`.
+Using the quick start example above as a reference:
```
Comment (helfi92) [inline README.md:227]: Correct me if I'm wrong, but the user will need to develop his components under `src/components`. Is it mentioned anywhere?

File: README.md [hunk 3]
```diff
@@ -217,6 +292,21 @@ module.exports = {
...
+    ['neutrino-preset-mozilla-rpweb', {
+      cssModules: true
+    }]
+  ]
+};
+```
+
 ## Hot Module Replacement
 
 While `neutrino-preset-react` supports Hot Module Replacement your app using React Hot Loader, it does require some
```
Comment (helfi92) [inline README.md:312]: s/your app using/for your app/

File: README.md [hunk 4]
```diff
@@ -265,6 +355,92 @@ if (module.hot) {
...
+The `<Stories />` component is the container for how a series of components should be rendered. It is responsible
+for rendering the navigation menu, switching between components and component states, and rendering the selected
+component.
+
+The `<Stories />` component should be given 1 or more `<Story />` components as children.
+
+```js
+import React from 'react';
+import { render } from 'react-dom';
+import { Stories } from 'neutrino-preset-react-component/lib';
```
Comment (helfi92) [inline README.md:374]: `import { Stories } from 'neutrino-preset-mozilla-rpweb/lib';`

**Review by helfi92 — CHANGES_REQUESTED** [review]
Looking forward for this to be merged!

File: src/index.js [hunk 5]
```diff
@@ -0,0 +1,216 @@
...
+    }
+  });
+
+  if (options.eslint) {
+    neutrino.use(loaderMerge('lint', 'eslint'), options.eslint);
+  }
+
+  if (options.cssModules) {
+    neutrino.config.module
+      .rule('plain-style')
```
Comment (edmorley) [inline src/index.js:106]: What's the difference between this and the stock Neutrino behaviour? Could you document it in a comment?
Should it be upstreamed to neutrino-middleware-style-loader?

File: src/index.js [hunk 6]
```diff
@@ -0,0 +1,216 @@
...
+
+  neutrino.config.resolve.modules.add(MODULES);
+  neutrino.config.resolveLoader.modules.add(MODULES);
+
+  neutrino.use(airbnb, {
+    eslint: {
+      plugins: ['eslint-plugin-react', 'eslint-plugin-prettier'],
+      baseConfig: {
+        extends: [
+          'plugin:react/recommended',
```
Comment (edmorley) [inline src/index.js:39]: IMO the react rules should be inherited from the airbnb react ruleset, and ideally Neutrino should do this for us. Have filed:
https://github.com/mozilla-neutrino/neutrino-dev/issues/352

File: src/index.js [hunk 7]
```diff
@@ -0,0 +1,216 @@
...
+      }
+    } catch (ex) {} // eslint-disable-line
+  }
+
+  neutrino.config.resolve.modules.add(MODULES);
+  neutrino.config.resolveLoader.modules.add(MODULES);
+
+  neutrino.use(airbnb, {
+    eslint: {
+      plugins: ['eslint-plugin-react', 'eslint-plugin-prettier'],
```
Comment (edmorley) [inline src/index.js:36]: I don't think the react plugin needs to be listed? Since it's set here:
https://github.com/mozilla-neutrino/neutrino-dev/blob/v7.1.0/packages/neutrino-preset-react/index.js#L40

File: src/index.js [hunk 8]
```diff
@@ -0,0 +1,216 @@
...
+        extends: [
+          'plugin:react/recommended',
+          'eslint-config-prettier',
+          'eslint-config-prettier/react'
+        ]
+      },
+      rules: {
+        // There are external dependencies we sometimes can't control, and
+        // renaming their variable to lowercase would not be idiomatic
+        'babel/new-cap': 'off',
```
Comment (edmorley) [inline src/index.js:47]: IMO we should either upstream these to Neutrino, here:
https://github.com/mozilla-neutrino/neutrino-dev/blob/v7.1.0/packages/neutrino-preset-airbnb-base/index.js#L11-L20

...or else stick to defaults.

File: src/index.js [hunk 9]
```diff
@@ -0,0 +1,216 @@
...
+    const components = join(neutrino.options.source, 'components');
+
+    neutrino.config.when(
+      neutrino.options.command === 'start',
+      () => {
+        neutrino.options.entry = 'stories'; // eslint-disable-line no-param-reassign
+      },
+      () => {
+        neutrino.config.entryPoints.delete('index');
+        readdirSync(components).map(component =>
```
Comment (edmorley) [inline src/index.js:175]: This seems like another use-case for Neutrino having multi-entrypoint support. (I've been meaning to file a Neutrino issue with a proposal for a Treeherder use case that would also simplify this.)

File: src/index.js [hunk 10]
```diff
@@ -0,0 +1,216 @@
...
+        (pkg.devDependencies && 'source-map-support' in pkg.devDependencies);
+
+      if (hasSourceMap) {
+        neutrino.use(banner);
+      }
+    } catch (ex) {} // eslint-disable-line
+  }
+
+  neutrino.config.resolve.modules.add(MODULES);
+  neutrino.config.resolveLoader.modules.add(MODULES);
```
Comment (edmorley) [inline src/index.js:32]: Why is this needed? Could you add a comment? Should it be upstreamed to Neutrino?

File: src/index.js [hunk 11]
```diff
@@ -0,0 +1,216 @@
...
+    (options.components || options.library) &&
+    neutrino.options.output.endsWith('build')
+  ) {
+    neutrino.options.output = 'lib'; // eslint-disable-line no-param-reassign
+
+    try {
+      const pkg = require(join(neutrino.options.root, 'package.json')); // eslint-disable-line
+      const hasSourceMap =
+        (pkg.dependencies && 'source-map-support' in pkg.dependencies) ||
+        (pkg.devDependencies && 'source-map-support' in pkg.devDependencies);
```
Comment (edmorley) [inline src/index.js:23]: Should this be upstreamed to Neutrino?

File: src/index.js [hunk 12]
```diff
@@ -0,0 +1,216 @@
...
+
+  if (options.envs) {
+    neutrino.config.plugin('env').tap(envs => [...envs, ...options.envs]);
+  }
+
+  neutrino.config.module
+    .rule('compile')
+    .use('cache')
+    .before('babel')
+    .loader(require.resolve('cache-loader'));
```
Comment (edmorley) [inline src/index.js:163]: What does this do? Could you add a comment? Should it be upstreamed?

File: src/index.js [hunk 13]
```diff
@@ -0,0 +1,216 @@
...
+    );
+  }
+
+  if (options.library || options.components) {
+    const name =
+      typeof options.library === 'string' ? options.library : '[name]';
+
+    neutrino.config
+      .devtool('source-map')
+      .performance.hints('warning')
```
Comment (edmorley) [inline src/index.js:190]: Could you add a comment explaining why this is being set? (I'm presuming it's since libraries are expected to be large)

File: src/index.js [hunk 14]
```diff
@@ -0,0 +1,216 @@
...
+
+    neutrino.config.when(process.env.NODE_ENV !== 'test', config => {
+      config.plugins
+        .delete('html')
+        .end()
+        .externals([nodeExternals()]);
+    });
+
+    neutrino.config.when(
+      neutrino.config.plugins.has('runtime-chunk'),
```
Comment (edmorley) [inline src/index.js:205]: I wonder if there should be an easier way to disable these plugins upsteam?

**Review by edmorley — COMMENTED** [review]
Some great cleanup and ideas in this project!

My only concerns are that:
* perhaps some of the functionality belongs in upstream Neutrino
* for the rest, it's a careful line between re-usable features and another large layer of abstraction that makes it quite hard to understand what's going on. Though hopefully with some more comments and some things upstreamed this can become less of an issue?
